File: mytower/game/views/renderers/elevator_renderer.py
# Copyright (c) 2025 Ryan Osterday. All rights reserved.
# See LICENSE file for details.

from __future__ import annotations  # Defer type evaluation

from typing import TYPE_CHECKING

import pygame

from mytower.game.core.units import Pixels, rect_from_pixels

if TYPE_CHECKING:
    from pygame import Surface

    from mytower.game.core.config import ElevatorCosmeticsProtocol
    from mytower.game.models.model_snapshots import ElevatorSnapshot
    from mytower.game.utilities.logger import LoggerProvider, MyTowerLogger


class ElevatorRenderer:

    def __init__(self, logger_provider: LoggerProvider, cosmetics_config: ElevatorCosmeticsProtocol) -> None:
        self._logger: MyTowerLogger = logger_provider.get_logger("ElevatorRenderer")
        self._cosmetics_config: ElevatorCosmeticsProtocol = cosmetics_config


    def draw(self, surface: Surface, elevator: ElevatorSnapshot) -> None:
        screen_height: Pixels = Pixels(surface.get_height())
        elevator_height: Pixels = self._cosmetics_config.ELEVATOR_HEIGHT.in_pixels

        elevator_top_z: Pixels = elevator.vertical_position.in_pixels
        elevator_top_y: Pixels = screen_height - elevator_top_z

        elevator_width: Pixels = self._cosmetics_config.ELEVATOR_WIDTH.in_pixels
        elevator_left_x: Pixels = elevator.horizontal_position.in_pixels

        color = self._cosmetics_config.OPEN_COLOR if elevator.door_open else self._cosmetics_config.CLOSED_COLOR

        pygame.draw.rect(
            surface, color, rect_from_pixels(elevator_left_x, elevator_top_y, elevator_width, elevator_height)
        )

File: web/src/rendering/ElevatorRenderer.ts
// Copyright (c) 2025 Ryan Osterday. All rights reserved.
// See LICENSE file for details.

/**
 * ElevatorRenderer - Responsible for drawing elevators
 * Matches Python's ElevatorRenderer
 */

import { Blocks } from '../units/Blocks';
import { PIXELS_PER_BLOCK } from './constants';
import { CoordinateTransform } from './CoordinateTransform';
import type { ElevatorSnapshotGQL } from '../generated/graphql';

export class ElevatorRenderer {
  private transform: CoordinateTransform;
  
  // Elevator dimensions (should match Python config)
  private readonly ELEVATOR_HEIGHT = 1.0;  // blocks
  private readonly ELEVATOR_WIDTH = 1.0;   // blocks

  constructor(
    private context: CanvasRenderingContext2D,
    canvasHeight: number
  ) {
    this.transform = new CoordinateTransform(canvasHeight);
  }

  public drawElevator(elevator: ElevatorSnapshotGQL): void {
    // World space (z): elevator position in blocks
    const elevatorTopZ = Blocks.from(elevator.verticalPosition);
    
    // Screen space (y): convert to pixels
    const elevatorTopY = this.transform.worldToScreen(elevatorTopZ);
    const elevatorLeftX = elevator.horizontalPosition * PIXELS_PER_BLOCK;
    
    // Dimensions
    const width = this.ELEVATOR_WIDTH * PIXELS_PER_BLOCK;
    const height = this.ELEVATOR_HEIGHT * PIXELS_PER_BLOCK;

    // Draw elevator car
    this.drawElevatorCar(elevatorLeftX, elevatorTopY.value, width, height, elevator.doorOpen);
    
    // Draw passenger count if any
    if (elevator.passengerCount > 0) {
      this.drawPassengerCount(elevatorLeftX, elevatorTopY.value, width, height, elevator.passengerCount);
    }
  }

  private drawElevatorCar(
    x: number,
    y: number,
    width: number,
    height: number,
    doorOpen: boolean
  ): void {
    const color = doorOpen ? '#c8c850' : '#3232c8';
    this.context.fillStyle = color;
    this.context.fillRect(x, y, width, height);
  }

  private drawPassengerCount(
    x: number,
    y: number,
    width: number,
    height: number,
    count: number
  ): void {
    this.context.fillStyle = '#ffffff';
    this.context.font = '12px monospace';
    this.context.textAlign = 'center';
    this.context.textBaseline = 'middle';
    this.context.fillText(
      `${count}`,
      x + width / 2,
      y + height / 2
    );
  }
}